gnocchi/amqp1d.py

```diff
@@ -0,0 +1,221 @@
+
+            names = list(measures_by_names.keys())
+            for name in names:
+                if name not in archive_policies:
+                    archive_policies[name] = (
+                        self.indexer.get_archive_policy_for_metric(name))
+            known_metrics = self.indexer.list_metrics(attribute_filter={
+                "and": [{"=": {"resource_id": resource.id}},
+                        {"in": {"name": names}}]
+            })
+            known_names = [m.name for m in known_metrics]
+            already_exists_names = []
+            for name in names:
+                if name not in known_names:
+                    try:
+                        m = self.indexer.create_metric(
+                            uuid.uuid4(),
+                            creator=self.conf.amqp1d.creator,
+                            resource_id=resource.id,
+                            name=name,
+                            archive_policy_name=archive_policies[name].name)
+                    except indexer.NamedMetricAlreadyExists as e:
+                        already_exists_names.append(e.metric)
+                    except indexer.IndexerException as e:
+                        LOG.error("Unexpected error, dropping metric %s",
+                                  name, exc_info=True)
+                    else:
+                        known_metrics.append(m)
+
+            if already_exists_names:
+                # Add metrics created in the meantime
+                known_names.extend(already_exists_names)
+                known_metrics.extend(
+                    self.indexer.list_metrics(
+                        names=already_exists_names,
+                        resource_id=resource.id)
+                )
+
+            self.incoming.add_measures_batch(
+                dict((metric.id,
+                     measures_by_names[metric.name])
+                     for metric in known_metrics))
+
+    def _get_resource(self, host_id):
+        resource_id = utils.ResourceUUID(host_id,
+                                         self.conf.amqp1d.creator)
+        resource = self.indexer.get_resource(self.conf.amqp1d.resource_type,
+                                             resource_id, with_metrics=True)
+        if resource:
+            return resource
+
+        try:
+            self.indexer.create_resource(self.conf.amqp1d.resource_type,
+                                         resource_id,
+                                         self.conf.amqp1d.creator,
+                                         original_resource_id=host_id,
+                                         host=self._hosts[host_id])
+        except indexer.ResourceAlreadyExists:
+            pass
+        else:
+            LOG.info("Created resource for %s", resource_id)
+
+        return self.indexer.get_resource(self.conf.amqp1d.resource_type,
+                                         resource_id, with_metrics=True)
```

Comment:
…no need to do a get if you just created it

Reply:
fixed